pennylane/compiler.py

```diff
@@ -0,0 +1,227 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This subpackage provides the JIT compilation support for PennyLane.
+
+
+This subpackage exclusively functions as a wrapper for PennyLane's JIT compiler packages, without independently
+implementing any compiler itself. Currently, it supports the ``pennylane-catalyst`` package, with plans to
+incorporate additional packages in the near future.
+
+For any compiler packages seeking to be registered, it is imperative that they expose the 'entry_points' metadata
+under the designated group name: ``pennylane.compilers``.
+"""
+
+from typing import List
+from importlib import reload
+import pkg_resources
+
+
+class Compiler:
+    """The JIT compiler module containing all utilities and wrapper methods for ``qml.qjit`` in PennyLane."""
+
+    # Private class properties
+    __BACKENDS = []
+    __ENTRY_POINTS = {}
+
+    @classmethod
+    def available(cls, name: str = "pennylane-catalyst") -> bool:
+        """Check the availability of the given compiler package.
+
+        Args:
+            name (str): the name of the compiler package
+
+        Return:
+            bool : True if the compiler package is installed on the system
+        """
+
+        reload(pkg_resources)
+        pkg_name = pkg_resources.Requirement.parse(name)
+
+        if pkg_resources.working_set.find(pkg_name):
+            if name not in cls.__BACKENDS:
+                cls.__ENTRY_POINTS = {
+                    entry.name: entry
+                    for entry in pkg_resources.iter_entry_points("pennylane.compilers")
+                }
+                cls.__BACKENDS.append(name)
+            return True
+
+        if name in cls.__BACKENDS:
+            cls.__BACKENDS.remove(name)
+
+        return False
+
+    @classmethod
+    def available_backends(cls) -> List[str]:
+        """Return the available compiler packages."""
+        return cls.__BACKENDS
+
+    @classmethod
+    def active(cls) -> bool:
+        """Check whether the caller is inside a QJIT evaluation context.
+
+        Return:
+            bool : True if the caller is inside a QJIT evaluation context
+        """
+        if not cls.__BACKENDS:
+            raise RuntimeError("There is no available compiler package.")
+
+        if "cpl_utils" not in cls.__ENTRY_POINTS:
+            raise RuntimeError("There is no available 'cpl_utils' entry point.")
+
+        utils_loader = cls.__ENTRY_POINTS["cpl_utils"].load()
+        return utils_loader.contexts.EvaluationContext.is_tracing()
```

Comment:
I think this needs to be generalized somehow so that it can check if any of the available compilers is active, right now it only checks one compiler.

Reply:
Just a thought, but since you mentioned in the standup that the entry point names couldn't be the same for different packages, maybe we check for a specific prefix/suffix in the available entry points?
So something like `catalyst_utils`, and `cuda_utils`, would be defined by the compiler packages, and we loop here looking for `*_utils`?
Just a thought off the top of my head.